ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/InsertValuesExecutor.java

```diff
@@ -0,0 +1,252 @@
+            key,
+            value
+        )
+    );
+
+    producer.close(Duration.ofSeconds(MAX_SEND_TIMEOUT_SECONDS));
+
+    return Optional.empty();
+  }
+
+  private GenericRow extractRow(
+      final InsertValues insertValues,
+      final StructuredDataSource<?> dataSource
+  ) {
+    final Optional<String> keyField = dataSource.getKeyField().name();
+
+    final List<String> columns = insertValues.getColumns().isEmpty()
+        ? dataSource.getSchema()
+          .fields()
+          .stream()
+          .map(Field::name)
+          .filter(name -> !SchemaUtil.ROWTIME_NAME.equals(name))
+          .collect(Collectors.toList())
+        : insertValues.getColumns();
+
+    if (columns.size() != insertValues.getValues().size()) {
+      // this will only happen if insertValues.getColumns() is empty, otherwise
+      // the columns/values are verified by the InsertValues constructor
+      throw new KsqlException(
+          "Expected a value for each column. Expected Columns: " + columns
+              + ". Got " + insertValues.getValues());
+    }
+
+    final Map<String, Object> values = new HashMap<>();
+    for (int i = 0; i < columns.size(); i++) {
+      final String column = columns.get(i);
+      final Schema columnSchema = dataSource.getSchema().field(column).schema();
+      final Expression valueExp = insertValues.getValues().get(i);
+
+      values.put(column, new ExpressionResolver(columnSchema, column).process(valueExp, null));
+    }
+
+    if (keyField.isPresent()) {
+      final String key = keyField.get();
+      final Object keyValue = values.get(key);
+      final Object rowKeyValue = values.get(SchemaUtil.ROWKEY_NAME);
+
+      if ((keyValue != null) ^ (rowKeyValue != null)) {
+        values.putIfAbsent(key, rowKeyValue);
+        values.putIfAbsent(SchemaUtil.ROWKEY_NAME, keyValue);
+      } else if (!Objects.equals(keyValue, rowKeyValue)) {
+        throw new KsqlException(
+            String.format(
+                "Expected ROWKEY and %s to match but got %s and %s respectively.",
+                key, rowKeyValue, keyValue));
+      }
+    }
+
+    values.putIfAbsent(SchemaUtil.ROWTIME_NAME, clock.getAsLong());
+
+    for (Field field : dataSource.getSchema().fields()) {
+      if (!field.schema().isOptional() && values.getOrDefault(field.name(), null) == null) {
+        throw new KsqlException("Got null value for nonnull field: " + field);
+      }
+    }
+
+    return new GenericRow(
+        dataSource.getSchema()
+            .fields()
+            .stream()
+            .map(Field::name)
+            .map(values::get)
+            .collect(Collectors.toList())
+    );
+  }
+
+  @SuppressWarnings("unchecked") // we know that key is String
+  private byte[] serializeKey(final Object keyValue, final StructuredDataSource<?> dataSource) {
+    if (keyValue == null) {
+      return null;
+    }
+
+    return ((Serde<String>) dataSource.getKeySerdeFactory().create())
```

Comment:
Do we validate the key is serializable before posting this command to the command topic?  Just because it can be serialized once doesn't mean that as KSQL evolves we don't end up with a data that will no longer serialize, i.e. its possible for this to fail.

I'd still wrap this in a try/catch and log to the processing log that we couldn't serialize the key (or value). This has the benefit that the user can configure if the bad key itself should be logged etc.

Reply:
See above, I believe that any errors should be returned to the user directly instead of logged. I will add a test for this.

Reply:
Yep - agree.

However, still would benefit from a re-try loop.